tools/apply-mapping/main.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+
+from argparse import ArgumentParser
+import zipfile
+import shutil
+import os
+import random
+import string
+import wkw
+import numpy as np
+import re
+import json
+
+def main():
+    args = create_parser().parse_args()
+
+    mapping = read_mapping(args)
+
+    tracing_tmpdir_path = extract_tracing_zip(args)
+    tracing_dataset = wkw.Dataset.open(os.path.join(os.path.join(tracing_tmpdir_path, 'data_zip'), '1'))
+    tracing_bboxes = file_bboxes(tracing_dataset)
+
+    print("Found {} tracing files".format(len(tracing_bboxes)))
+
+    def apply_maping_to_voxel(voxel):
+        voxel_str = str(voxel)
+        if voxel_str in mapping:
+            return np.uint32(mapping[voxel_str])
+        else:
+            return voxel
+
+    def is_mapped(voxel):
+        return str(voxel) in mapping
+
+    changed_count = 0
+
+    for counter, tracing_bbox in enumerate(tracing_bboxes):
+        print("Processing tracing file {} of {}, bounding box: {}...".format(counter+1, len(tracing_bboxes), tracing_bbox))
+        data = tracing_dataset.read(tracing_bbox[0], tracing_bbox[1])
+        transformed = np.vectorize(apply_maping_to_voxel)(data)
+        tracing_dataset.write(tracing_bbox[0], transformed)
+        changed_count += np.count_nonzero(np.vectorize(is_mapped)(data))
+
+    print("Changed {} Voxels".format(changed_count))
+    pack_tracing_zip(tracing_tmpdir_path, args.tracing_path)
+    print("Done")
+
+def read_mapping(args):
+    if not (args.mapping_path or args.mapping_str):
+        raise Exception("""No mapping supplied, please add a mapping string in the format '{"7":"3", "12":"3"}' or a mapping file path via -f my_mapping.json""")
+
+    if (args.mapping_path and args.mapping_str):
+        print("Warning: both mapping string and mapping file path were supplied, ignoring the file.")
+
+    if (args.mapping_str):
+        mapping = json.loads(args.mapping_str)
+    else:
+        with open(args.mapping_path) as mapping_file:
+            mapping = json.load(mapping_file)
+    print("Using mapping: ", mapping)
+    return mapping
+
+def extract_tracing_zip(args):
+    tracing_tmpdir_path = tmp_filename()
+    os.makedirs(tracing_tmpdir_path)
+    with zipfile.ZipFile(args.tracing_path) as outer_zip:
+        zipfile.ZipFile.extractall(outer_zip, path=tracing_tmpdir_path)
+    with zipfile.ZipFile(os.path.join(tracing_tmpdir_path, 'data.zip')) as data_zip:
+        zipfile.ZipFile.extractall(data_zip, path=os.path.join(tracing_tmpdir_path, 'data_zip'))
+    return tracing_tmpdir_path
+
+def pack_tracing_zip(tracing_tmpdir_path, tracing_path):
+    os.remove(os.path.join(tracing_tmpdir_path, 'data.zip'))
```

Comment:
If the input data is named 'data.zip' this will make an in-place-replacement, right? Would this work?